### DISCUSSION TITLE
How to deal with Spread spectrum induced polarization data using Pygmili?

### DISCUSSION THREAD

**Discussion by kouabena** [gimli-org/pyGIMLi#748] [General]
I'm dealing spread Spectrum induced polarization data and I want to use Pygmili to do my inversion

**Comment by halbmy** [comment]
A discussion should provide some background information and raise questions or open points, otherwise nobody will join.

Please have a look at the SIP examples on https://www.pygimli.org

**Comment by kouabena** [comment]
Ok you are right Sir, Indeed, I am using spread spectrum-induced
polarization data and would like to know the possibility of using pyGimli
for inversion and my Ph.D thesis

Le mar. 9 juil. 2024 à 15:21, Thomas Günther ***@***.***> a
écrit :

**Comment by halbmy** [comment]
There is not more information contained. Are you working on (lab) SIP spectroscopy or on frequency-domain field IP?

In the first case, you will find some examples on https://www.pygimli.org , in the latter there is a dedicated class in the BERT project.

**Comment by kouabena** [comment]
I'm working on frequency-domain field IP

**Comment by kouabena** [comment]
My dataset is different from the normal SIP and IP, which is why I would like to know if there is a window in Pygmili for my dataset.
I have passed through the website, but I cannot see anything suitable for my dataset. I am willing to share the dataset with you so you can take a look and let me know if Pygmili can help me with my thesis.

**Comment by halbmy** [comment]
Yes, for analysing frequency-domain field-scale spectral IP data you should consider the `FDIP` class in the BERT repository. Details are outlined by Günther & Martin (2016) and Martin et al. (2020) along with example codes on https://zenodo.org/records/4419736

References
* Günther, T. & Martin, T. (2016): Spectral two-dimensional inversion of frequency-domain induced polarisation data from a mining slag heap. Journal of Applied Geophysics 135, 436-448, doi:10.1016/j.jappgeo.2016.01.008.
* Martin, T., Günther, T., Orozco, A.F. & Dahlin, T. (2020): Evaluation of spectral induced polarization field measurements in time and frequency domain, J. Appl. Geophys. 180, 104141, doi:10.1016/j.jappgeo.2020.104141.

**Comment by kouabena** [comment]
Good day Sir , I'm working with Spread spectrum induced polarization data
incversion using pyGIMLI . Indeed, I would like to receive guidance from
you and see how I can effectively apply pyGIMLI for my PhD thesis . You can
find in attached the code and the data . I'm looking forward to hearing
from you.
Best regards.
Wilfried

Le mer. 28 août 2024 à 14:15, Thomas Günther ***@***.***> a
écrit :

**Comment by halbmy** [comment]
I have not yet understood what exactly you want to do. Did you read the papers I've mentioned? Do you want to invert 2D (or 3D) FDIP field data in a spectral manner or do you want to analyse laboratory spectra?

**Comment by kouabena** [comment]
I want to invert 2D and 3D field FDIP data in a spectra manner . I read the
papers and my data configuration looks different that is why I keep asking.



Le mer. 14 mai 2025 à 01:29, Thomas Günther ***@***.***> a
écrit :

**Comment by halbmy** [comment]
The inversion is independent on the specific arrays (quadrupoles) used. A sketch or example file would be needed to give you some guideline.

Please consider opening an issue in the [BERT repository](https://gitlab.com/resistivity-net/bert)

**Comment by kouabena** [comment]
Thank you, I will do that.

Le mer. 14 mai 2025 à 01:40, Thomas Günther ***@***.***> a
écrit :

**Comment by kouabena** [comment]
<img width="1168" height="559" alt="image" src="https://github.com/user-attachments/assets/ecffe36d-e1e3-4456-8e55-798ffeeb5301" />
please am running inversion with Pygimli and I'm getting this error when I want to load the dataset . Could you help me handling it

**Reply by halbmy** [reply]
Can you please attach the file and also specify which pygimli version you're using?

**Comment by kouabena** [comment]
[SIP_inv_F2_src_1.ipynb](https://github.com/user-attachments/files/22245847/SIP_inv_F2_src_1.ipynb)
The version is 1.5.4

**Comment by halbmy** [comment]
I meant the data file. We can't fix the problem if we can't reproduce it.

**Comment by kouabena** [comment]
[L22_F1.zip](https://github.com/user-attachments/files/22247884/L22_F1.zip)
here is the data

**Comment by halbmy** [comment]
In the data section of res2dinv files (general array), the first number is the number of electrodes
```
...
3.000000  -20.000000  87.211998   40.000000  89.773003   80.000000  95.232002   89.519066   4.475953     8.000743   10.000000
3.000000  -20.000000  87.211998   80.000000  95.232002  120.000000 103.751999   70.552818   3.527641    -2.670331   10.000000
3.000000  -20.000000  87.211998  120.000000 103.751999  160.000000 106.464996   67.367546   3.368377    22.828770   10.000000
...
```
You wrote it like a float value causing the error (like the error message states). If you replace it by `3`, all works well. Nevertheless I changed the import routine so that `3.00000` will also be read.

**Reply by kouabena** [reply]
you mean you already changed the import routine? or not yet and I should remove .000000 ?

**Comment by halbmy** [comment]
I've already changed the import routine but you would have to git-checkout the source code to use it. So it is maybe easier (and more rigorous) to Search-and-Replace (or change format in Excel) in the data file.

**Reply by kouabena** [reply]
<img width="1138" height="373" alt="image" src="https://github.com/user-attachments/assets/78ce3789-ff93-4e37-a16b-5d268715033a" />
still getting the same error after replacing 3.0000 to 3   
[cleaned_L22-F1-src1.zip](https://github.com/user-attachments/files/22268405/cleaned_L22-F1-src1.zip)

**Reply by kouabena** [reply]
<img width="1138" height="373" alt="image" src="https://github.com/user-attachments/assets/78ce3789-ff93-4e37-a16b-5d268715033a" />
still getting the same error after replacing 3.0000 to 3   
[cleaned_L22-F1-src1.zip](https://github.com/user-attachments/files/22268405/cleaned_L22-F1-src1.zip)

**Reply by halbmy** [reply]
Can you upload the changed data file again? I am afraid you did a mistake as the code tries to convert `Error` of line 13 into an integer.

**Reply by kouabena** [reply]
here is it 
[cleaned_L22-F1-src1.zip](https://github.com/user-attachments/files/22268615/cleaned_L22-F1-src1.zip)

**Comment by kouabena** [comment]
I just started the dev option and it seems to work

**Comment by kouabena** [comment]
it works but not with the 3.0000 so I used the changed dataset

**Comment by halbmy** [comment]
The current `dev` branch works with both files, I guess there was another bug fixed since v1.5.4.
So it is really time to draft a new version.

**Comment by kouabena** [comment]
I finished my first test , could you take a look at it and make if necessarly some suggestions to improve the code and the results. 
[SSIP_inv_code_L22-F1_SRC1.py](https://github.com/user-attachments/files/22273353/SSIP_inv_code_L22-F1_SRC1.py)

**Reply by kouabena** [reply]
hello Sir please did you take a look at the file I sent you ?

**Comment by halbmy** [comment]
I did and it looks generally ok (a few things can be simplified) but as the data only contain one source, I cannot assess whether the inversion makes sense

**Comment by kouabena** [comment]
Here is the full dataset containing four frequencies (F1-F4) .
[Uploading L22_F1.zip…]()

**Reply by halbmy** [reply]
this is not a link to any data. If it comes to analysing multi-frequency spectrally, you should (maybe after filtering the data separately) switch to the FDIP class in the pyBERT module.

**Reply by kouabena** [reply]
[L22_F1.zip](https://github.com/user-attachments/files/22356640/L22_F1.zip)
PLlease can you check this file

**Reply by halbmy** [reply]
the data look ok but contain very low apparent resistivity that should probably be removed before inversion. However, hand-inversion is difficult as the data set is really huge. Try filtering and inverting each data set separately before going into spectral analysis.

**Reply by kouabena** [reply]
ok got it

**Reply by kouabena** [reply]
hello , please I tried to run the entire data , but sounds like it is not working well instead of taking sources.
When I pick one source from the data and run the inversion it gives a result but when I pick the entire data it gives an error

**Reply by halbmy** [reply]
Can you paste the error message? Otherwise we won't be able to help you.

**Reply by kouabena** [reply]
<img width="1129" height="717" alt="image" src="https://github.com/user-attachments/assets/84929662-ef56-4ec6-bff2-40b17d500197" />

**Comment by kouabena** [comment]
Hello Sir please did you take a look at the error

**Comment by kouabena** [comment]
hello sir please try to take a look at the error am getting running the inversion

**Comment by halbmy** [comment]
I just tried 
```
data = ert.load("L22_F2.dat")
data.remove(data1["rhoa"] < 30)
data.estimateError()
mgr = ert.ERTIPManager(data)
mgr.invertDC(maxIter=0, verbose=True)
mgr.invertIP(maxIter=1, verbose=True)
```
and it works well, but I didn't try the others.
From the error I guess you have zero IP values.

**Comment by kouabena** [comment]
here is the script line that gives me error. 
<img width="1184" height="563" alt="image" src="https://github.com/user-attachments/assets/607e6c33-ca6b-4fc0-b4cf-99fddcfc78b4" />

**Comment by halbmy** [comment]
`invert()` only calls `invertDC()` and `invertIP()`, so please check whether you have the same error message for the two lines I suggested and whether it occurs for all data files or just for some of them.

**Comment by kouabena** [comment]
it is working now following what you gave above, but not satisfied by the results. The resistivity is very low.
<img width="1182" height="792" alt="image" src="https://github.com/user-attachments/assets/2505c907-a383-4faf-a380-037d78705f3a" />

**Comment by halbmy** [comment]
Of course you have to remove the `maxIter` (maximum iteration number) argument. I just used it to test it without spending much runtime.

**Reply by kouabena** [reply]
and there is no any value for lam? or I have to run the script you gave me while removing only the maxIter

**Comment by kouabena** [comment]
here is the first output after removing the maxIter 
<img width="668" height="698" alt="image" src="https://github.com/user-attachments/assets/70514150-f3d6-438b-936b-e539703ff13e" />

**Comment by halbmy** [comment]
Good it is working technically. In order to achieve good results, you should probably look at the data and model response and the misfit (I just extended `mgr.showFit()` to IP) and work on better filtering.

**Comment by kouabena** [comment]
ok Sir thank you , I will try and let you know. It is my Ph.d research data and I would like to publish a paper with pyGIMLI

**Comment by kouabena** [comment]
HERE IS THE FIT i GOT 
<img width="815" height="528" alt="image" src="https://github.com/user-attachments/assets/f8e124a4-b932-4c6a-8cac-a9c4ccfaa629" />
